File: src/main/java/xyz/oribuin/flighttrails/manager/DataManager.java
package xyz.oribuin.flighttrails.manager;

import org.bukkit.Bukkit;
import org.bukkit.Particle;
import org.bukkit.block.data.BlockData;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.ItemStack;
import xyz.oribuin.flighttrails.FlightTrails;
import xyz.oribuin.flighttrails.data.PlayerData;
import xyz.oribuin.flighttrails.util.FileUtils;
import xyz.oribuin.flighttrails.util.TrailColor;

import java.io.File;
import java.io.IOException;
import java.util.*;

/**
 * @author Esophose
 */

public class DataManager extends Manager implements Listener {

    // Define data.yml
    private final static String DATA_CONFIG = "data.yml";

    private FileConfiguration dataConfig;
    private final Map<UUID, PlayerData> playerData;

    public DataManager(FlightTrails plugin) {
        super(plugin);
        this.playerData = new HashMap<>();

        // Register the events
        Bukkit.getPluginManager().registerEvents(this, this.plugin);
    }

    @Override
    public void reload() {
        // Check file
        FileUtils.createFile(this.plugin, DATA_CONFIG);

        // Load configuration
        this.dataConfig = YamlConfiguration.loadConfiguration(this.getDataFile());

        // Clear PlayerData HashMap
        this.playerData.clear();

        // Load all players data
        for (Player player : Bukkit.getOnlinePlayers())
            this.getPlayerData(player, false);
    }

    // Load PlayerData on PlayerJoin
    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        this.getPlayerData(event.getPlayer(), false);
    }

    // Remove the player from hashMap when they leave
    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        this.playerData.remove(event.getPlayer().getUniqueId());
    }

    // Get the player's Data
    public PlayerData getPlayerData(Player player, boolean create) {
        // Get the PlayerData from hashmap
        PlayerData playerData = this.playerData.get(player.getUniqueId());

        // Check if PlayerData is null
        if (playerData == null) {

            // Define PlayerUUID
            String key = player.getUniqueId().toString();

            // Get the Player from data.yml
            ConfigurationSection playerDataSection = this.dataConfig.getConfigurationSection(key);
            if (playerDataSection == null) {
                if (!create)
                    return null;

                // Get default values
                playerData = PlayerData.getDefault(player);

                // Save default values
                this.savePlayerData(playerData);
            } else {
                playerData = new PlayerData(player, playerDataSection);
            }

            // Add playerdata to HashMap
            this.playerData.put(player.getUniqueId(), playerData);
        }

        // Return PlayerData
        return playerData;
    }

    // Get all PlayerData
    public Collection<PlayerData> getAllPlayerData() {
        return Collections.unmodifiableCollection(this.playerData.values());
    }

    /**
     * Save all the PlayerData inside data.yml
     *
     * @param playerData the Data being saved
     */
    private void savePlayerData(PlayerData playerData) {
        ConfigurationSection dataSection = this.getPlayerDataSection(playerData.getPlayer());

        // Set Values
        dataSection.set("enabled", playerData.isEnabled());
        dataSection.set("particle", playerData.getParticle().name());
        dataSection.set("item", playerData.getItem().getType().name());
        dataSection.set("block", playerData.getBlock().getMaterial().name());
        dataSection.set("color", playerData.getTrailColor().name());

        // Save values
        this.saveData();
    }

    // Get the Player from data.yml
    private ConfigurationSection getPlayerDataSection(Player player) {
        // Get Player UUID
        String key = player.getUniqueId().toString();

        // If Player is saved, get Data
        if (this.dataConfig.isConfigurationSection(key))
            return this.dataConfig.getConfigurationSection(key);

        // If player does not exist, create it
        return this.dataConfig.createSection(key);
    }

    public void updatePlayerData(Player player, boolean enabled) {
        // Define PlayerData
        PlayerData playerData = this.getPlayerData(player, true);

        // Set Trails as enabled
        playerData.setEnabled(enabled);

        // Set enabled to true
        this.getPlayerDataSection(player).set("enabled", enabled);

        // Save Data
        this.saveData();
    }

    public void updatePlayerData(Player player, Particle particle) {
        // Define PlayerData
        PlayerData playerData = this.getPlayerData(player, true);

        // Set the particle
        playerData.setParticle(particle);

        // Set particle in configuration section
        this.getPlayerDataSection(player).set("particle", particle.name());

        // Save Data
        this.saveData();
    }

    public void updatePlayerData(Player player, ItemStack item) {
        // Define PlayerData
        PlayerData playerData = this.getPlayerData(player, true);

        // Set the ItemStack
        playerData.setItem(item);

        // Set the ItemStack inside configuration section
        this.getPlayerDataSection(player).set("item", item.getType().name());

        // Save Data
        this.saveData();
    }

    public void updatePlayerData(Player player, BlockData block) {
        // Define PlayerData
        PlayerData playerData = this.getPlayerData(player, true);

        // Set the block
        playerData.setBlock(block);

        // Set the block in the configuration section
        this.getPlayerDataSection(player).set("block", block.getMaterial().name());

        // Save Data
        this.saveData();
    }

    public void updatePlayerData(Player player, TrailColor color) {
        // Define PlayerData
        PlayerData playerData = this.getPlayerData(player, true);

        // Set TrailColor
        playerData.setTrailColor(color);

        // Set TrailColor in the configuration section
        this.getPlayerDataSection(player).set("color", color.name());

        // Save Data
        this.saveData();
    }

    private void saveData() {
        try {
            this.dataConfig.save(this.getDataFile());
        } catch (IOException ex) {
            ex.printStackTrace();
        }
    }

    private File getDataFile() {
        return new File(this.plugin.getDataFolder(), DATA_CONFIG);
    }

}

File: src/main/java/xyz/oribuin/flighttrails/manager/DataManager.kt
package xyz.oribuin.flighttrails.manager

import org.bukkit.Color
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import xyz.oribuin.flighttrails.FlightTrails
import xyz.oribuin.orilibrary.util.FileUtils.createFile
import xyz.oribuin.orilibrary.manager.Manager
import java.io.File

/**
 * @author Oribuin
 */
class DataManager(plugin: FlightTrails) : Manager(plugin) {
    private var dataConfig: FileConfiguration? = null

    // Reload Method
    override fun enable() {
        createFile(plugin, "data.yml")
        dataConfig = YamlConfiguration.loadConfiguration(dataFile)
    }

    // Disable Method
    override fun disable() {
        // Unused
    }

    /**
     * Get or set the trail particle for a player
     *
     * @param player The player being set.
     * @param particle The particle being set.
     */
    fun getOrSetParticle(player: Player, particle: Particle?): Particle {
        if (particle != null) {
            getConfig().set(player.uniqueId.toString() + ".particle", particle.name)
            saveData()
            return particle
        }

        return Particle.valueOf(getConfig().getString(player.uniqueId.toString() + ".particle")?.toUpperCase() ?: return Particle.valueOf(plugin.config.getString("particle-settings.default.particle") ?: return Particle.FLAME))
    }

    /**
     * Get or set the trail block for a player
     *
     * @param player The player being set.
     * @param material The material being set.
     */
    fun getOrSetBlock(player: Player, material: Material?): Material {
        if (material != null) {
            getConfig().set(player.uniqueId.toString() + ".block", material.name)
            saveData()
            return material
        }

        return Material.valueOf(getConfig().getString(player.uniqueId.toString() + ".block")?.toUpperCase() ?: return Material.valueOf(plugin.config.getString("particle-settings.default.block") ?: return Material.LIGHT_BLUE_WOOL))
    }

    /**
     * Get or set the trail Item for a player
     *
     * @param player The player being set.
     * @param itemStack The item being set.
     */
    fun getOrSetItem(player: Player, itemStack: ItemStack?): ItemStack {
        if (itemStack != null) {
            getConfig().set(player.uniqueId.toString() + ".item", itemStack)
            saveData()
            return itemStack
        }

        return getConfig().getItemStack(player.uniqueId.toString() + ".item") ?: return plugin.config.getItemStack("particle-settings.default.item") ?: return ItemStack(Material.LIGHT_BLUE_WOOL)
    }

    /**
     * Get or set the trail color for a player
     *
     * @param player The player being set.
     * @param color The color
     * being set.
     */
    fun getOrSetColor(player: Player, color: Color?): Color {
        if (color != null) {
            getConfig().set(player.uniqueId.toString() + ".color", color)
            saveData()
            return color
        }

        return getConfig().getColor(player.uniqueId.toString() + ".color") ?: return plugin.config.getColor("particle-settings.default.color") ?: return Color.AQUA
    }

    /**
     * Get or set if the player has trails
     *
     * @param player The player being set
     * @param enabled if they are enabled or not
     */
    fun getOrSetEnabled(player: Player, enabled: Boolean?): Boolean {
        if (enabled != null) {
            getConfig().set(player.uniqueId.toString() + ".enabled", enabled)
            saveData()
            return enabled
        }

        return getConfig().getBoolean(player.uniqueId.toString() + ".enabled")
    }

    /**
     * Set the default values for a player
     */
    fun setDefault(player: Player) {
        val uuid = player.uniqueId.toString()
        getConfig().set("$uuid.enabled", plugin.config.getBoolean("particle-settings.default.enabled"))
        getConfig().set("$uuid.block", getOrSetBlock(player, null).name)
        getConfig().set("$uuid.color", getOrSetColor(player, null))
        getConfig().set("$uuid.particle", getOrSetParticle(player, null).name)
        getConfig().set("$uuid.item", getOrSetItem(player, null))
        saveData()
    }

    /**
     * Get the player from the config
     *
     * @param player The player being obtained
     */
    fun getPlayer(player: Player): Player? {
        return getConfig().getOfflinePlayer(player.uniqueId.toString())?.player
    }

    /**
     * Get the configuration as not nullable
     *
     * @return dataConfig
     */
    fun getConfig(): FileConfiguration {
        return dataConfig ?: return plugin.config
    }

    /**
     * Save the trail data
     */
    private fun saveData() {
        dataConfig?.save(dataFile)
    }

    /**
     * Get the Data File from plugin folder
     */
    private val dataFile: File get() = File(plugin.dataFolder, "data.yml")
}